server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -1111,6 +1111,20 @@ public boolean canOperateOnVolume(Volume volume) {
     }
 
     @Override
+    public void cleanupSnapshotsByVolume(Long volumeId) {
+        List<SnapshotVO> volSnapShots = _snapshotDao.listByVolumeId(volumeId);
+        for(SnapshotVO snapshot: volSnapShots) {
+            SnapshotInfo info = snapshotFactory.getSnapshot(snapshot.getId(), DataStoreRole.Primary);
```

Comment:
This appears to be an application side join.  Please consider creating a new query to retrieve all snapshot info instances associated with `volumeId` to reduce load on the database and simplify this method.